src/emqx_protocol.erl

```diff
@@ -777,20 +792,22 @@ shutdown(Reason, #pstate{client_id = ClientId}) when Reason =:= conflict;
     emqx_cm:unregister_connection(ClientId);
 shutdown(Reason, PState = #pstate{connected = true,
                                   client_id = ClientId,
-                                  will_msg  = WillMsg}) ->
+                                  will_msg  = WillMsg,
+                                  session   = Session}) ->
     ?LOG(info, "Shutdown for ~p", [Reason], PState),
-    _ = send_willmsg(WillMsg),
+    _ = send_willmsg(WillMsg, Session),
     emqx_hooks:run('client.disconnected', [credentials(PState), Reason]),
     emqx_cm:unregister_connection(ClientId).
 
-send_willmsg(undefined) ->
+send_willmsg(undefined, _Session) ->
     ignore;
 send_willmsg(WillMsg = #message{topic = Topic,
-                                headers = #{'Will-Delay-Interval' := Interval}})
+                                headers = #{'Will-Delay-Interval' := Interval}}, Session)
             when is_integer(Interval), Interval > 0 ->
     SendAfter = integer_to_binary(Interval),
-    emqx_broker:publish(WillMsg#message{topic = <<"$delayed/", SendAfter/binary, "/", Topic/binary>>});
-send_willmsg(WillMsg) ->
+    Session1 = list_to_binary(pid_to_list(Session)),
```

Comment:
Question: why would we need to have the pid in the topic ?
Pid tend to be ephemeral because it may crash/restart, make sense to have the client-id instead ?

Reply:
Yes, you are right, client id will be better

Reply:
I can't seem to find the code (in this review diff) which tries to find the session back and delete something.